### PR TITLE
feat: support http range header

### PR DESCRIPTION
**Pull request by SgtPooki**
## Title

Add Range header support.

## Description

Fixes https://github.com/ipfs/helia-verified-fetch/issues/9

## Notes & open questions

- [x] Still need to add some additional tests, but it's ready for code review.
- [x] byte-range-context.spec.ts is currently skipping `handling ReadableStreams` tests, but these are covered by `range requests` tests.. going to try to fix quickly and if I can't i'll remove them.

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if necessary (this includes comments as well)
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: packages/verified-fetch/src/utils/responses.ts
Comment (SgtPooki) [inline packages/verified-fetch/src/utils/responses.ts:46]: not sure if we want this to be `badInternalResponse` or other, so i went verbose

File: packages/verified-fetch/src/verified-fetch.ts
Comment (SgtPooki) [inline packages/verified-fetch/src/verified-fetch.ts:304]: I'm not sure if we care about offset and length here, but I presumed so. If not, we can remove a lot of this code up here and potentially remove the `eslint-disable-next-line complexity`

File: packages/verified-fetch/src/verified-fetch.ts
Comment (SgtPooki) [inline packages/verified-fetch/src/verified-fetch.ts:424]: returning an invalid response 500 response here resulted in needing to update `custom-dns-resolvers.spec.ts`. We may want to provide more detailed information than a 500 error here. Maybe a 503 "service unavailable" ?

File: packages/verified-fetch/test/range-requests.spec.ts
Comment (SgtPooki) [inline packages/verified-fetch/test/range-requests.spec.ts:66]: I need to implement range-requests for 'raw' and add raw tests here

File: packages/verified-fetch/src/verified-fetch.ts
Comment (SgtPooki) [inline packages/verified-fetch/src/verified-fetch.ts:471]: I was able to move out some of the code from the handleDagPb, but could probably move this function outside of `verified-fetch` altogether.

**Review by SgtPooki — COMMENTED** [review]
self review, some comments

File: packages/verified-fetch/src/utils/request-headers.ts
Comment (achingbrain) [inline packages/verified-fetch/src/utils/request-headers.ts:57]: Don't know if we couldn't simplify this regex with something like:

```js
/^bytes=(?<start>\d+)?-(?<end>\d+)?$/
```

Then guard on neither `match.start` or `match.end` being present?

File: packages/verified-fetch/src/utils/responses.ts
Comment (achingbrain) [reply packages/verified-fetch/src/utils/responses.ts:46]: I think call it what it is, 500 is `Internal Server Error` so that's what we should call it.

By that logic `notSupportedResponse` should also be `notImplementedResponse`, but whatevs.

File: packages/verified-fetch/src/utils/responses.ts
Comment (achingbrain) [inline packages/verified-fetch/src/utils/responses.ts:134]: It's probably worth logging the error to help ourselves in the future?

File: packages/verified-fetch/src/verified-fetch.ts
Comment (achingbrain) [inline packages/verified-fetch/src/verified-fetch.ts:362]: Can you open an issue or better yet, a PR with some failing test cases for when `unixfs.cat` isn't throwing when it should?

File: packages/verified-fetch/src/verified-fetch.ts
Comment (achingbrain) [inline packages/verified-fetch/src/verified-fetch.ts:420]: Just a tiny consistency thing, all of our errors are thrown with messages in sentence case, logging is all done in lower case.  Not sure why, but it's where we are.

```suggestion
      this.log.error('error streaming %c/%s', cid, path, err)
```

File: packages/verified-fetch/src/verified-fetch.ts
Comment (achingbrain) [reply packages/verified-fetch/src/verified-fetch.ts:471]: It could be made a function in the util folder?  No strong opinion.

File: packages/verified-fetch/src/utils/request-headers.ts
Comment (achingbrain) [inline packages/verified-fetch/src/utils/request-headers.ts:48]: Can we simplify the input types here and just make it `Headers`?

File: packages/verified-fetch/src/utils/request-headers.ts
Comment (achingbrain) [inline packages/verified-fetch/src/utils/request-headers.ts:5]: This is only used in the `getRequestRange` function, no need to export it.

If the input to `getRequestRange` is [simplified to Headers] this function can be removed entirely?

File: packages/verified-fetch/src/utils/response-headers.ts
Comment (achingbrain) [inline packages/verified-fetch/src/utils/response-headers.ts:7]: Minor nit but synchronous APIs normally have "sync" appended, not prepended.  For example the [fs sync API](https://nodejs.org/api/fs.html#synchronous-api).

```suggestion
function getBodySizeSync (body: SupportedBodyTypes): number | null {
```

**Comment by achingbrain** [comment]
> ive got a refactor going that cleans this up a bunch

Done a first pass, will look again after the refactor lands.

File: packages/verified-fetch/src/utils/request-headers.ts
Comment (SgtPooki) [reply packages/verified-fetch/src/utils/request-headers.ts:57]: yea I was debating doing that. I like the explicitness of what is already there, but I'm good with changing it. I'm sure having less alternatives will make it quicker. will give an attempt at this after I finish up the byte-range refactor

File: packages/verified-fetch/src/verified-fetch.ts
Comment (SgtPooki) [reply packages/verified-fetch/src/verified-fetch.ts:362]: Yep. I wanted to make sure I was actually passing the right things.. i've got the test expectations and params cognizable now, so I'm sure something is off with .cat. will open issues when I can

File: packages/verified-fetch/test/custom-dns-resolvers.spec.ts
Comment (SgtPooki) [reply packages/verified-fetch/test/custom-dns-resolvers.spec.ts:31]: I felt the same about a 500 error. The closest error I could think of was 503, 409, 411 or some others depending on why exactly the streaming fails

File: packages/verified-fetch/src/utils/request-headers.ts
Comment (SgtPooki) [reply packages/verified-fetch/src/utils/request-headers.ts:48]: I was trying to normalize the request Header processing and response header generation further with the existing code.

Unfortunately, the code I have now is already hard for *me* to adjust (and too many changes needed in too many places), and I wrote it, so I started creating a `byte-range-context.ts` file with a class `ByteRangeContext` that should be able to handle things more easily. When that's done, much of the byte-range handling should be much more straightforward.

Right now, calling getRequestRange, and getContentRangeHeader, and all the possible response values require too much context within handleDagPb & handleRaw, so I think this will get much better

File: packages/verified-fetch/src/utils/request-headers.ts
Comment (SgtPooki) [reply packages/verified-fetch/src/utils/request-headers.ts:5]: I think we will continue to have a need for a getHeader function throughout this application though?

File: packages/verified-fetch/src/verified-fetch.ts
Comment (SgtPooki) [reply packages/verified-fetch/src/verified-fetch.ts:375]: I talked with some folks earlier this week on this and it would make sense to default to 502 Bad Gateway

File: packages/verified-fetch/src/verified-fetch.ts
Comment (SgtPooki) [reply packages/verified-fetch/src/verified-fetch.ts]: this function goes away with the addition of the byte-range-context class

File: packages/verified-fetch/src/verified-fetch.ts
Comment (SgtPooki) [reply packages/verified-fetch/src/verified-fetch.ts]: fixed in [`f357a3d` (#10)](https://github.com/ipfs/helia-verified-fetch/pull/10/commits/f357a3d5eaa240a4808c8461cce65619fba2f1e4)

File: packages/verified-fetch/src/verified-fetch.ts
Comment (SgtPooki) [reply packages/verified-fetch/src/verified-fetch.ts]: apparently it's working fine, it was a bug in my code :) surprised?! haha

File: packages/verified-fetch/test/custom-dns-resolvers.spec.ts
Comment (SgtPooki) [reply packages/verified-fetch/test/custom-dns-resolvers.spec.ts]: discussed with @lidel the other day, kubo used to throw 502.. we can use that for now. fixed in [`83e80d8` (#10)](https://github.com/ipfs/helia-verified-fetch/pull/10/commits/83e80d85839cff64d5f9e8be308a0bf1d963382b)

File: packages/verified-fetch/src/utils/request-headers.ts
Comment (SgtPooki) [reply packages/verified-fetch/src/utils/request-headers.ts]: function is removed and ByteRangeContext now handles parsing of header info upon construction.

File: packages/verified-fetch/src/utils/request-headers.ts
Comment (SgtPooki) [reply packages/verified-fetch/src/utils/request-headers.ts:1]: resolving this as I think it makes sense for a normalized header getting function to exist in a request-headers utility file

File: packages/verified-fetch/src/utils/responses.ts
Comment (SgtPooki) [reply packages/verified-fetch/src/utils/responses.ts:129]: done in [`05a6dfb` (#10)](https://github.com/ipfs/helia-verified-fetch/pull/10/commits/05a6dfb90631b527e078a4f9c61954c0a66e7905)

File: packages/verified-fetch/src/verified-fetch.ts
Comment (SgtPooki) [reply packages/verified-fetch/src/verified-fetch.ts]: removed the offset & length here.

File: packages/verified-fetch/src/utils/response-headers.ts
Comment (SgtPooki) [inline packages/verified-fetch/src/utils/response-headers.ts:3]:
```suggestion
 * If you know the total size of the body, pass it as `byteSize`
```

File: packages/verified-fetch/src/utils/response-headers.ts
Comment (SgtPooki) [inline packages/verified-fetch/src/utils/response-headers.ts:21]:
```suggestion
  if (byteStart != null && byteEnd == null) {
    if (byteSize == null) {
      return `bytes */${total}`
    }
    return `bytes ${byteStart}-${byteSize}/${byteSize}`
  }
  if (byteStart == null && byteEnd != null) {
    if (byteSize == null) {
      return `bytes */${total}`
    }
    return `bytes ${byteSize - byteEnd + 1}-${byteSize}/${byteSize}`
  }
```

File: packages/verified-fetch/src/verified-fetch.ts
Comment (SgtPooki) [inline packages/verified-fetch/src/verified-fetch.ts:342]:
```suggestion
    // we have to .stat before .cat so we can get the size and make sure byte offsets are calculated properly
    if (byteRangeContext.isRangeRequest && byteRangeContext.isValidRangeRequest) {
      stat = stat ?? await this.unixfs.stat(resolvedCID, {
        signal: options?.signal,
        onProgress: options?.onProgress
      })
      this.log.trace('stat.fileSize for rangeRequest %d', stat.fileSize)
      byteRangeContext.fileSize = stat.fileSize
    }
```

File: packages/verified-fetch/test/range-requests.spec.ts
Comment (SgtPooki) [inline packages/verified-fetch/test/range-requests.spec.ts:32]:
```suggestion
```

File: packages/verified-fetch/test/range-requests.spec.ts
Comment (SgtPooki) [inline packages/verified-fetch/test/range-requests.spec.ts:52]:
```suggestion
```

File: packages/verified-fetch/test/utils/byte-range-context.spec.ts
Comment (SgtPooki) [inline packages/verified-fetch/test/utils/byte-range-context.spec.ts:100]: currently investigating unskipping/removing these

**Review by SgtPooki — COMMENTED** [review]
self review

**Review by SgtPooki — COMMENTED** [review]
self review

File: packages/verified-fetch/test/utils/byte-range-context.spec.ts
Comment (SgtPooki) [reply packages/verified-fetch/test/utils/byte-range-context.spec.ts]: unskipped.

File: packages/verified-fetch/src/utils/byte-range-context.ts
Comment (achingbrain) [inline packages/verified-fetch/src/utils/byte-range-context.ts:112]:
```suggestion
    this.log.error('returning unmodified body for valid range request')
```

File: packages/verified-fetch/src/utils/byte-range-context.ts
Comment (achingbrain) [inline packages/verified-fetch/src/utils/byte-range-context.ts:271]: If we're going to use jsdoc comments, the comment block needs to be `/** .... */` and not `//`

```suggestion
  /**
   * This function returns the value of the "content-range" header.
   * Content-Range: <unit> <range-start>-<range-end>/<size>
   * Content-Range: <unit> <range-start>-<range-end>/*
   * Content-Range: <unit> */<size>
   * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Range
   */
```

File: packages/verified-fetch/src/utils/request-headers.ts
Comment (achingbrain) [inline packages/verified-fetch/src/utils/request-headers.ts:6]: The `?? undefined` is redundant?

```suggestion
    return headers.get(header)
```

File: packages/verified-fetch/src/types.ts
Comment (achingbrain) [inline packages/verified-fetch/src/types.ts:3]: Picking nits but `Types` is redundant in a type, `Supported` is a business-logic kind of decision not a type, so just `Body`?

```suggestion
export type Body = string | ArrayBuffer | Blob | ReadableStream<Uint8Array> | null
```

File: packages/verified-fetch/src/verified-fetch.ts
Comment (achingbrain) [inline packages/verified-fetch/src/verified-fetch.ts:337]: This stat operation is unnecessary - you can disambiguate the `terminalElement` type by examining the `.type` field:

```js
if (terminalElement.type === 'file') {
  // tsc now knows terminalElement is `UnixFSFile`
  byteRangeContext.fileSize = terminalElement.unixfs.fileSize()
}

```

File: packages/verified-fetch/src/utils/byte-range-context.ts
Comment (achingbrain) [inline packages/verified-fetch/src/utils/byte-range-context.ts:119]: All this casting removes any type safety we might have.  Would it be better to convert any passed body types to just one type instead of having to deal with all of: `Uint8Array | ArrayBuffer | Blob | string`?

Convert everything to `Blob` or something?

File: packages/verified-fetch/src/utils/byte-range-context.ts
Comment (achingbrain) [inline packages/verified-fetch/src/utils/byte-range-context.ts:150]: If this class is just used internally, do we have to protect ourselves from ourselves with read-only properties like this?

File: packages/verified-fetch/src/utils/byte-range-context.ts
Comment (SgtPooki) [reply packages/verified-fetch/src/utils/byte-range-context.ts:271]: +1 this was leftover from copying and pasting from MDN and doing an auto comment with vscode

File: packages/verified-fetch/src/types.ts
Comment (SgtPooki) [reply packages/verified-fetch/src/types.ts:3]: I would rather do something like `ResponseBody`, but i'm good with any.

`Body` could easily be a type that comes from builtin/global types that could cause confusion. `SupportedBody` would be better I guess.. but it's explicitly informing devs reading the code that it's not just typical `Response.body` types.

**Review by SgtPooki — COMMENTED** [review]
self review and response to comments

**Comment by SgtPooki** [comment]
FYI: some tests were failing when merging main into byte-range PR due to new ipns things. changes at [`acdd632` (#10)](https://github.com/ipfs/helia-verified-fetch/pull/10/commits/acdd632437d9e5914e0d5a10adf1905c2e67ce04)

File: packages/verified-fetch/src/utils/byte-range-context.ts
Comment (achingbrain) [reply packages/verified-fetch/src/utils/byte-range-context.ts]: Ok, it makes sense for `fileSize`, that change should probably be reverted.  What about `_isRangeRequest`?  It's a `readonly` boolean that's returned from a getter for `isRangeRequest`.

My concern with getters/setters is that they usually start off with good intentions around data access, but the temptation is always there to do business-logic things in them, which means accessing properties can mutate state which in the long term can make codebases very hard to reason about.

File: packages/verified-fetch/src/types.ts
Comment (achingbrain) [reply packages/verified-fetch/src/types.ts:3]: `ResponseBody` would be fine.  It's a minor point tbh.

File: packages/verified-fetch/src/utils/byte-range-context.ts
Comment (achingbrain) [inline packages/verified-fetch/src/utils/byte-range-context.ts:53]: `SupportedBodyTypes` includes `null`

```suggestion
  private _body: SupportedBodyTypes = null
```

**Review by achingbrain — APPROVED** [review]
LGTM, pending remaining discussion points

File: packages/verified-fetch/src/utils/byte-range-context.ts
Comment (SgtPooki) [inline packages/verified-fetch/src/utils/byte-range-context.ts:289]:
```suggestion
   * - Content-Range: <unit> <byteStart>-<byteEnd>/*
```

Comment (SgtPooki) [reply packages/verified-fetch/src/utils/byte-range-context.ts:289]: I was testing jsdoc workarounds for js comment block closing tags and forgot to remove

File: packages/verified-fetch/src/utils/byte-range-context.ts
Comment (SgtPooki) [inline packages/verified-fetch/src/utils/byte-range-context.ts:291]: @achingbrain is this better than all as single line comments?

File: packages/verified-fetch/src/verified-fetch.ts
Comment (SgtPooki) [inline packages/verified-fetch/src/verified-fetch.ts:293]:
```suggestion
    let stat: UnixFSStats
```

Can probably remove this again since we only call stat once. If it requires no changes ill impl, otherwise ill leave it

File: packages/verified-fetch/src/verified-fetch.ts
Comment (SgtPooki) [inline packages/verified-fetch/src/verified-fetch.ts:331]:
```suggestion
    // if terminalElement is a file, we know the size
```

File: packages/verified-fetch/src/verified-fetch.ts
Comment (SgtPooki) [inline packages/verified-fetch/src/verified-fetch.ts:453]: Less throwing, more bad fetch responses

File: packages/verified-fetch/src/utils/byte-range-context.ts
Comment (SgtPooki) [inline packages/verified-fetch/src/utils/byte-range-context.ts:54]:
```suggestion
  private readonly rangeRequestHeader: string | undefined
```

File: packages/verified-fetch/src/utils/byte-range-context.ts
Comment (SgtPooki) [inline packages/verified-fetch/src/utils/byte-range-context.ts:68]:
```suggestion
      this.isRangeRequest = true
```

**Review by SgtPooki — COMMENTED** [review]
Going over code one last time

File: packages/verified-fetch/src/utils/byte-range-context.ts
Comment (SgtPooki) [reply packages/verified-fetch/src/utils/byte-range-context.ts]: done

**Comment by SgtPooki** [comment]
commenting elsewhere since the thread (https://github.com/ipfs/helia-verified-fetch/pull/10#discussion_r1521644018) was collapsed:

we could solve both concerns by setting internalFileSize so we know which value we set vs consumers.. but that could get confusing. i'll leave as is for now.

i've got _isRangeRequest updated locally, and have isValidRangeRequest updated as well.

So now we have:

inaccessable properties outside the ByteRangeContext class:

1. `_fileSize` // setter and getter
2. `_contentRangeHeaderValue` // getter (depends on other operations)
3. `_body` // set/get methods
4. ~~`_isValidRangeRequest` // getter (depends on other operations)~~

the only outlier here seems to be _body having `setBody` and `getBody` methods

---

inline update: `_isValidRangeRequest` has been removed. see [`d633456` (#10)](https://github.com/ipfs/helia-verified-fetch/pull/10/commits/d6334569dcf33ac63098ee8890ad3caab1114d3b)

**Pull request merged**
